### PR TITLE
Move from DeviceID to TargetID in data placement

### PR DESCRIPTION
**Pull request by ChristopherHogan**
* Changed `PlacementSchema` from `std::vector<std::pair<size_t, DeviceID>>` to `std::vector<std::pair<size_t, TargetID>>`.
* All data placement now works with `TargetID`s instead of `DeviceID`s.
* I made some name changes in the DPE code to help improve my understanding. Let me know if I got anything wrong.
* We still only work with local targets, but this PR provides the infrastructure to support neighborhoods and global lists of targets.

### CODE REVIEW

File: src/hermes_types.h [hunk 1]
```diff
@@ -46,11 +46,22 @@ constexpr char kPlaceInHierarchy[] = "PlaceInHierarchy";
 #define HERMES_NOT_IMPLEMENTED_YET \
   LOG(FATAL) << __func__ << " not implemented yet\n"
 
+union TargetID {
+  struct {
```
Comment (jya-kmu) [inline src/hermes_types.h:50]: Can we have an explanation here how to understand this struct as a `TargetID`?

**Review by jya-kmu — APPROVED** [review]
Looks good to me. Thanks!

File: src/hermes_types.h (hunk 1 above)
Comment (ChristopherHogan) [reply src/hermes_types.h:51]: I added some documentation. Does that answer your question?

Comment (jya-kmu) [reply src/hermes_types.h:51]: Yes, thanks!

**Pull request merged**